# Lariat Waveforms 

Adam Lyon / Fermilab SCD / September 2015

!! Not Complete !!

Jim has code to output the waveform information from a Lariat event. Let's look at it in 3D in ParaView.

The data is in the form of a big CSV file

In [1]:
import numpy as np
import pandas as pd

In [52]:
di = pd.read_csv("data20150919/output_di.csv", header=None)

In [53]:
di.head()

0            1     2     3        4     5     6     7     8     9     ...   \
0     1  SlicerInput   239  3072  2156.59     1     3     3     3     4  ...    
1     1  SlicerInput   238  3072  2157.98     1     0     1     2     3  ...    
2     1  SlicerInput   237  3072  2153.17     1     6     6     7     8  ...    
3     1  SlicerInput   236  3072  2122.89     1    -3     0     2     0  ...    
4     1  SlicerInput   235  3072  2146.26     1    -1    -1    -1    -1  ...    

   3068  3069  3070  3071  3072  3073  3074  3075  3076  3077  
0    -2    -2    -1    -1     0     0     1     1     2     2  
1    -2    -2    -2    -1    -1     0     1     3     4     5  
2     0     1     1     1     1     3     4     3     4     4  
3     0    -1    -3    -3     1     2    -1     0     2     0  
4    -1    -1     0     1     0     0     1     1     0     1  

[5 rows x 3078 columns]

Columns `0,1` and `3:5` may be removed. Column 2 is the wire number. The ADC counts start at column 6. 

In [54]:
di.drop(di.columns[0:2], axis=1, inplace=True)
di.drop(di.columns[1:4], axis=1, inplace=True)

In [55]:
di.head()

2     6     7     8     9     10    11    12    13    14    ...   3068  \
0   239     3     3     3     4     4     4     4     4     4  ...     -2   
1   238     0     1     2     3     3     5     6     6     6  ...     -2   
2   237     6     6     7     8     7     7     7     7     6  ...      0   
3   236    -3     0     2     0    -2     1     2    -1    -1  ...      0   
4   235    -1    -1    -1    -1    -1     0    -1    -1     0  ...     -1   

   3069  3070  3071  3072  3073  3074  3075  3076  3077  
0    -2    -1    -1     0     0     1     1     2     2  
1    -2    -2    -1    -1     0     1     3     4     5  
2     1     1     1     1     3     4     3     4     4  
3    -1    -3    -3     1     2    -1     0     2     0  
4    -1     0     1     0     0     1     1     0     1  

[5 rows x 3073 columns]

In [56]:
di.shape

(480, 3073)

Let's sort by the wire number so that things come out in the right order. 

In [58]:
di = di.sort_values([2])
di.head()

2     6     7     8     9     10    11    12    13    14    ...   3068  \
239     0     2     1     1     1     1     1     1     0     0  ...      3   
238     1     1     0     1     1     1     0     0     1     0  ...      2   
237     2     2     1     1     1     0    -1    -1    -2    -1  ...      0   
236     3     0    -1     0     0     0    -1    -1     0    -1  ...      5   
235     4     2     1     1     1     1     1     1     1     1  ...      3   

     3069  3070  3071  3072  3073  3074  3075  3076  3077  
239     3     4     4     3     4     4     4     4     4  
238     2     3     3     4     3     4     4     3     4  
237     1     2     2     1     2     1     1     2     1  
236     5     6     7     7     7     7     7     7     6  
235     3     3     3     3     3     3     4     3     3  

[5 rows x 3073 columns]

There are 480 wires in the event. 240 "U" and 240 "V". Column `2` is the wire number. Wires # < 240 are U and wire # >= 240 are V. 

Let's get the U wires

In [59]:
u = di[ di[2] < 240 ]

In [60]:
u.shape

(240, 3073)

In [61]:
u.head()

2     6     7     8     9     10    11    12    13    14    ...   3068  \
239     0     2     1     1     1     1     1     1     0     0  ...      3   
238     1     1     0     1     1     1     0     0     1     0  ...      2   
237     2     2     1     1     1     0    -1    -1    -2    -1  ...      0   
236     3     0    -1     0     0     0    -1    -1     0    -1  ...      5   
235     4     2     1     1     1     1     1     1     1     1  ...      3   

     3069  3070  3071  3072  3073  3074  3075  3076  3077  
239     3     4     4     3     4     4     4     4     4  
238     2     3     3     4     3     4     4     3     4  
237     1     2     2     1     2     1     1     2     1  
236     5     6     7     7     7     7     7     7     6  
235     3     3     3     3     3     3     4     3     3  

[5 rows x 3073 columns]

In [62]:
v = di[ di[2] >= 240]

In [65]:
v.head()

2     6     7     8     9     10    11    12    13    14    ...   3068  \
479   240     4     3     2     3     2     1     0     0    -1  ...      5   
478   241     2     2     2     2     1     1     1     1    -1  ...     -2   
477   242    11    10     9     8     7     6     6     4     4  ...      1   
476   243     6     5     5     4     3     3     2     2     2  ...      0   
475   244     1     1     0     0     0     0     0     0     0  ...     -1   

     3069  3070  3071  3072  3073  3074  3075  3076  3077  
479     6     7     7     7     8     8     8     9     8  
478    -2    -2    -2    -1     0     0     1     2     2  
477     1     1     2     4     3     4     4     4     5  
476     1     2     2     3     3     4     5     4     5  
475    -1     1     0     1     1     1     2     1     2  

[5 rows x 3073 columns]

In [66]:
v.shape

(240, 3073)

So there are 240 U wires, 240 V wires, and each has 3072 time entries. Each entry has an ampliude. The U and V values go together. What we eventually want is a structure like 

    U,V,T,Uamp,Vamp
    
Where we'll transform `U,V,T` into position values $x,y,z$ with two scalar arrays for the amplitudes. The "cell" is a wire crossing (e.g. U#1 crosses V#200 at a time T with amplitude Uamp on the U wire and Vamp on the V wire))

Let's relable the columns

In [67]:
u.columns = ['U'] + range(3072)
v.columns = ['V'] + range(3072)

In [68]:
u.head()

U  0  1  2  3  4  5  6  7  8  ...   3062  3063  3064  3065  3066  3067  \
239  0  2  1  1  1  1  1  1  0  0  ...      3     3     4     4     3     4   
238  1  1  0  1  1  1  0  0  1  0  ...      2     2     3     3     4     3   
237  2  2  1  1  1  0 -1 -1 -2 -1  ...      0     1     2     2     1     2   
236  3  0 -1  0  0  0 -1 -1  0 -1  ...      5     5     6     7     7     7   
235  4  2  1  1  1  1  1  1  1  1  ...      3     3     3     3     3     3   

     3068  3069  3070  3071  
239     4     4     4     4  
238     4     4     3     4  
237     1     1     2     1  
236     7     7     7     6  
235     3     4     3     3  

[5 rows x 3073 columns]

In [69]:
v.head()

V   0   1  2  3  4  5  6  7  8  ...   3062  3063  3064  3065  3066  \
479  240   4   3  2  3  2  1  0  0 -1  ...      5     6     7     7     7   
478  241   2   2  2  2  1  1  1  1 -1  ...     -2    -2    -2    -2    -1   
477  242  11  10  9  8  7  6  6  4  4  ...      1     1     1     2     4   
476  243   6   5  5  4  3  3  2  2  2  ...      0     1     2     2     3   
475  244   1   1  0  0  0  0  0  0  0  ...     -1    -1     1     0     1   

     3067  3068  3069  3070  3071  
479     8     8     8     9     8  
478     0     0     1     2     2  
477     3     4     4     4     5  
476     3     4     5     4     5  
475     1     1     2     1     2  

[5 rows x 3073 columns]

For the `V` wire #'s bring them back into a range of 0-239 to match the `U` wires

In [70]:
v.loc[:,'V'] = v.loc[:,'V']-240

In [71]:
v.head()

V   0   1  2  3  4  5  6  7  8  ...   3062  3063  3064  3065  3066  3067  \
479  0   4   3  2  3  2  1  0  0 -1  ...      5     6     7     7     7     8   
478  1   2   2  2  2  1  1  1  1 -1  ...     -2    -2    -2    -2    -1     0   
477  2  11  10  9  8  7  6  6  4  4  ...      1     1     1     2     4     3   
476  3   6   5  5  4  3  3  2  2  2  ...      0     1     2     2     3     3   
475  4   1   1  0  0  0  0  0  0  0  ...     -1    -1     1     0     1     1   

     3068  3069  3070  3071  
479     8     8     9     8  
478     0     1     2     2  
477     4     4     4     5  
476     4     5     4     5  
475     1     2     1     2  

[5 rows x 3073 columns]

Now we want to reshape data frames to be `U,T,amp` (replace `U` with `V` for the other data frame).

Let's write out VTK image files for the U and V wires (separately) so we can see the wave forms. We'll do $x$ as the time, $y$ as the wire number, and the $z$ will always be 1 (we'll either color or warp to see the amplitude). Note that I previously had the wire as $x$ but I think it will look better with $t$ going horizontally. Loading a VTK image means that the $x$ component iterates first, then $y$, then $z$. So this means that we want data in the form of,

    adcU0T0, adcV0T0
    adcU0T1, adcV0T1
    ...
    adcU0T3071, adc0T3071
    adcU1T0, adcV1T0
    ...
    
So we want to run down the columns and output
    

In [109]:
w = open('waveform.csv', 'w')
w.write("adcU/I,adcV/I\n")  # Attributes of the grid
w.write("%s,%s,1\n" % (u.shape[1]-1, u.shape[0]))  # dimensions  (need to drop the wire number column)
w.write("1,1,1\n")  # Spacing of 1
w.write("0,0,0\n")  # origin

In [113]:
# Writing out x=wire, y=time  Ok - this is really ugly
#for timeCol in range(u.shape[1]-1):
#    out = zip( u.iloc[:, timeCol+1], v.iloc[:, timeCol+1])  # Add one since first col is wire #
#    for i in range(len(out)):
#        w.write("%s,%s  # %s,%s\n" % (out[i][0], out[i][1], i, timeCol) )

In [111]:
# Writing out x=time, y=wire  OK - this is really ugly
for wireRow in range(u.shape[0]):
    out = zip( u.iloc[wireRow][1:], v.iloc[wireRow][1:] )  # Skip the wire number column
    for i in range(len(out)):
        w.write("%s,%s  # time=%s wire=%s\n" % (out[i][0], out[i][1], i, wireRow) )

In [112]:
w.close()

In [203]:
um = pd.melt(u, id_vars='U')
um.columns = ['U', 'T', 'Uamp']
um = um.sort_values(['T', 'U'])
um.head()

U  T  Uamp
239  0  0     2
238  1  0     1
237  2  0     2
236  3  0     0
235  4  0     2

In [204]:
vm = pd.melt(v, id_vars="V")
vm.columns = ['V','T','Vamp']
vm = vm.sort_values(['T', 'V'])
vm.head()

V  T  Vamp
239  0  0     4
238  1  0     2
237  2  0    11
236  3  0     6
235  4  0     1

In [205]:
um.shape

(737280, 3)

In [206]:
vm.shape

(737280, 3)

Top plot these for individual wires, make vti with extents 240,240,1 and write 1 for z. Then write the amp and warp by scalar

New we merge them (takes a few seconds)

In [224]:
d = pd.merge(um, vm, how='inner', on='T')

In [225]:
d.shape

(176947200, 5)

In [226]:
d.head()

U  T  Uamp  V  Vamp
0  0  0     2  0     4
1  0  0     2  1     2
2  0  0     2  2    11
3  0  0     2  3     6
4  0  0     2  4     1

In [227]:
d

U     T  Uamp    V  Vamp
0            0     0     2    0     4
1            0     0     2    1     2
2            0     0     2    2    11
3            0     0     2    3     6
4            0     0     2    4     1
5            0     0     2    5    -1
6            0     0     2    6     0
7            0     0     2    7     2
8            0     0     2    8    -1
9            0     0     2    9     0
10           0     0     2   10     2
11           0     0     2   11     3
12           0     0     2   12     1
13           0     0     2   13     2
14           0     0     2   14     3
15           0     0     2   15     3
16           0     0     2   16    -1
17           0     0     2   17     3
18           0     0     2   18     2
19           0     0     2   19     1
20           0     0     2   20     1
21           0     0     2   21     1
22           0     0     2   22    -2
23           0     0     2   23    -1
24           0     0     2   24    -2
25           0     0     2   25    -2
26           0     0     2   26    -3
27           0     0     2   27     0
28           0     0     2   28     2
29           0     0     2   29    -1
...        ...   ...   ...  ...   ...
176947170  239  3071     2  210     1
176947171  239  3071     2  211     0
176947172  239  3071     2  212    -2
176947173  239  3071     2  213    -2
176947174  239  3071     2  214     0
176947175  239  3071     2  215     0
176947176  239  3071     2  216     0
176947177  239  3071     2  217    -2
176947178  239  3071     2  218     0
176947179  239  3071     2  219    -1
176947180  239  3071     2  220    -2
176947181  239  3071     2  221    -1
176947182  239  3071     2  222     1
176947183  239  3071     2  223     3
176947184  239  3071     2  224     9
176947185  239  3071     2  225     3
176947186  239  3071     2  226     6
176947187  239  3071     2  227    -1
176947188  239  3071     2  228     7
176947189  239  3071     2  229     7
176947190  239  3071     2  230     7
176947191  239  3071     2  231     7
176947192  239  3071     2  232     1
176947193  239  3071     2  233     4
176947194  239  3071     2  234     1
176947195  239  3071     2  235     0
176947196  239  3071     2  236     1
176947197  239  3071     2  237     4
176947198  239  3071     2  238     3
176947199  239  3071     2  239     3

[176947200 rows x 5 columns]

In [228]:
d.shape

(176947200, 5)

In [229]:
240*240*3072

176947200

In [199]:
d.sort(['U','V','T'])

/Users/lyon/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  if __name__ == '__main__':


KeyboardInterrupt: 

Ok - this is **NOT** the way to do this. The sort is way too slow. Instead, just get the values directly from the dataframes and construct the reduced ordered dataframes needed to make the VTI